tlsutil/config.go

```diff
@@ -233,156 +234,186 @@ func (c *Configurator) commonTLSConfig(additionalVerifyIncomingFlag bool) (*tls.
 // otherwise we assume that no TLS should be used.
-func (c *Configurator) OutgoingRPCConfig() (*tls.Config, error) {
-	useTLS := c.base.CAFile != "" || c.base.CAPath != "" || c.base.VerifyOutgoing
-	if !useTLS {
-		return nil, nil
+func (c *Configurator) OutgoingRPCConfig() *tls.Config {
+	c.log("OutgoingRPCConfig")
+	c.RLock()
+	defer c.RUnlock()
+	if c.outgoingRPCTLSDisabled() {
+		return nil
 	}
 	return c.commonTLSConfig(false)
 }
 
+func (c *Configurator) outgoingRPCTLSDisabled() bool {
+	return c.base.CAFile == "" && c.base.CAPath == "" && !c.base.VerifyOutgoing
+}
+
 // OutgoingRPCWrapper wraps the result of OutgoingRPCConfig in a DCWrapper. It
 // decides if verify server hostname should be used.
 func (c *Configurator) OutgoingRPCWrapper() (DCWrapper, error) {
-	// Get the TLS config
-	tlsConfig, err := c.OutgoingRPCConfig()
-	if err != nil {
-		return nil, err
-	}
-
-	// Check if TLS is not enabled
-	if tlsConfig == nil {
+	c.log("OutgoingRPCWrapper")
+	if c.outgoingRPCTLSDisabled() {
 		return nil, nil
 	}
 
-	// Generate the wrapper based on hostname verification
+	// Generate the wrapper based on dc
 	wrapper := func(dc string, conn net.Conn) (net.Conn, error) {
-		if c.base.VerifyServerHostname {
-			// Strip the trailing '.' from the domain if any
-			domain := strings.TrimSuffix(c.base.Domain, ".")
-			tlsConfig = tlsConfig.Clone()
-			tlsConfig.ServerName = "server." + dc + "." + domain
-		}
-		return c.base.wrapTLSClient(conn, tlsConfig)
+		return c.wrapTLSClient(dc, conn)
 	}
 
 	return wrapper, nil
 }
 
-// AddCheck adds a check to the internal check map together with the skipVerify
-// value, which is used when generating a *tls.Config for this check.
-func (c *Configurator) AddCheck(id string, skipVerify bool) {
-	c.Lock()
-	defer c.Unlock()
-	c.checks[id] = skipVerify
+func (c *Configurator) log(name string) {
+	if c.logger != nil {
+		c.logger.Printf("[DEBUG] tlsutil: %s with version %d", name, c.version)
+	}
 }
 
-// RemoveCheck removes a check from the internal check map.
-func (c *Configurator) RemoveCheck(id string) {
-	c.Lock()
-	defer c.Unlock()
-	delete(c.checks, id)
-}
+// Wrap a net.Conn into a client tls connection, performing any
+// additional verification as needed.
+//
+// As of go 1.3, crypto/tls only supports either doing no certificate
+// verification, or doing full verification including of the peer's
+// DNS name. For consul, we want to validate that the certificate is
+// signed by a known CA, but because consul doesn't use DNS names for
+// node names, we don't verify the certificate DNS names. Since go 1.3
+// no longer supports this mode of operation, we have to do it
+// manually.
+func (c *Configurator) wrapTLSClient(dc string, conn net.Conn) (net.Conn, error) {
+	var err error
+	var tlsConn *tls.Conn
 
-func (c *Configurator) getSkipVerifyForCheck(id string) bool {
-	c.Lock()
-	defer c.Unlock()
-	return c.checks[id]
+	c.RLock()
+	config := c.OutgoingRPCConfig()
```

Comment:
Hmm `OutgoingRPCConfig` also takes an `RLock` on the same lock.

Multiple Read locks are allowed so this won't obvious deadlock and tests will pass, but there is a subtle bug here due to Go's RWMutex implementation that I happened to read about the other day.

See https://stackoverflow.com/questions/30547916/goroutine-blocks-when-calling-rwmutex-rlock-twice-after-an-rwmutex-unlock

> This means that it is always unsafe to call RLock on a RWMutex that the same goroutine already has read locked.

I don't know if `-race` is smart enough to spot that but we should avoid it here explicitly - maybe with an internal `outgoingRPCConfigLocked` implementation?

Reply:
great catch! `-race` is not smart enough. :(
I am restructuring the code a little so that all reading happens from independent small functions that can be locked without having to think about nested locking.